testament/testament.nim

```diff
@@ -1220,155 +1497,908 @@ proc parseArgs(execState: var Execution, p: var OptParser): ParseCliResult =
+    if cat.string notin ["testdata", "nimcache"]:
+      let catId = execState.categories.len
+      execState.categories.add cat
+
+      var handled = true
+
+      let normCat = cat.string.normalize
+      if isCompilerRepo:
+        case normCat
+        of "gc":
+          setupGcTests(execState, catId)
+        of "threads":
+          setupThreadTests(execState, catId)
+        of "lib":
+          # IMPLEMENT: implement this proc and all the subsequent handling
+          setupStdlibTests(execState, catId)
+        else:
+          handled = false
+
+      if not handled:
+        for file in walkDirRec(testsDir & cat.string):
+          if file.isTestFile:
+            execState.testFiles.add:
+              TestFile(file: file, catId: catId)
+
+  case filter.kind:
+    of tfkAll:
+      let testsDir = testsDir
+      for kind, dir in walkDir(testsDir):
+        if kind == pcDir:
+          # The category name is extracted from the directory
+          # eg: 'tests/compiler' -> 'compiler'
+          let cat = dir[testsDir.len .. ^1]
+          testFilesFromCat(execState, Category(cat))
+      if isCompilerRepo: # handle `AdditionalCategories`
+        for cat in AdditionalCategories:
+          testFilesFromCat(execState, Category(cat))
+
+    of tfkCategories:
+      for cat in filter.cats:
+        testFilesFromCat(execState, cat)
+
+    of tfkGlob:
+      execState.categories = @[Category "<glob>"]
+      let pattern = filter.pattern
+      if dirExists(pattern):
+        for kind, name in walkDir(pattern):
+          if kind in {pcFile, pcLinkToFile} and name.endsWith(".nim"):
+            execState.testFiles.add TestFile(file: name)
+      else:
+        for name in walkPattern(pattern):
+          execState.testFiles.add TestFile(file: name)
+
+    of tfkSingle:
+      execState.categories = @[Category parentDir(filter.test)]
+      let test = filter.test
+      # IMPLEMENT: replace with proper error handling
+      doAssert fileExists(test), test & " test does not exist"
+      if isTestFile(test):
+        execState.testFiles.add TestFile(file: test)
+
+    else:
+      assert false, "TODO ???"
+
+  execState.testFiles.sort # ensures we have a reproducible ordering
+
+  # parse all specs
+  for testId, test in pairs(execState.testFiles):
+    execState.testSpecs.add parseSpec(
+      addFileExt(test.file, ".nim"),
+      execState.categories[test.catId].defaultTargets(),
+      nativeTarget()
+    )
+
+    if execState.testOpts.hasKey(testId):
+      # apply additional test matrix, if specified
+      let optMatrix = execState.testOpts[testId].optMatrix
+      execState.testSpecs[testId].matrix =
+        case execState.testSpecs[testId].matrix.len
+        of 0:
+          optMatrix
+        else:
+          # (saem)REFACTOR - this is a hack, we shouldn't modify the
+          # spec.matrix
+
+          # (haxscramper)QUESTION Doesn't "specify additional test matrix
```

Comment:
The assignment  to `execState.testSpecs[testId].matrix` is a bad idea, we should keep the spec intact.

Ideally, a spec is just what was declared in the file, actually should be called `FileSpec`, `DeclaredSpec`, or something. Then we create actual specs that are truly atomic specs, so a declared spec should generate 1 or more actual specs given matrix options, target, etc declared data.

Reply:
this is an acceptable compromise for a good while, but it'll eventually run into limits when we want to support multiple specs per file, more complex matricies, and or more complex test steps/orchestration.

Reply:
Ok, so this is mostly an issue of "MatrixAsSpecfied" vs "MatrixAsActuallyUsed"; got it. `DeclaredSpec` seems like a good name for it